Add script to help people identify duplicate vaults

scottisloud opened the pull request:
This script creates a CSV file that may help customers identify duplicate vaults. Once they've identified the vaults they'd like deleted, they could pass a list of vault UUIDs to a second script (in progress) that would delete those vaults. 

Because there is no single `op` command that gives us all we need to make a csv with the useful information, it may be a bit API-intensive (e.g., we need to make separate calls for _each_ vault to `op vault user list` so we can get a count of users and `op vault get` to get details like number of items, date created). 

As such it's a bit slow, which isn't the end of the world, but I'm concerned it may flirt with our rate limits for a very large customer, however the slowness may help mitigate that. 

It's a bit of a monolith, but I figured it's easier for people to use a single .py file than to split it out as modules.

scottisloud commented:
Recent commits add the following:
* An "add_vault_prefix" companion script that accepts a newline-delimited list of vault UUIDS (such as that which you might extract from a csv of vaults created by the "vault_dedupe_helper" script. )
  * This script optionally allows a person to delete the vaults in the list. 
* Adds some error handling to both scripts so user will know which vaults raised errors. 
* Provide feedback to the user when running any of the scripts. 

Re the "add_vault_prefix" script: 
* This script requires a `--file` option which expects a file path for the list of vault UUIDs. 
* Currently the script runner must comment the `prefixVault()` function call and uncomment the `deleteVault()` call to delete vaults. 
  * This is deliberately clunky because I want to make it as difficult as possible to delete vaults. But is this the right choice? **Should we instead use an optional `-d` flag which, when provided by the user, deletes, rather than prefixes?**

ag-adampike commented on migration/clean-up-scripts/add_vault_prefix.py:
```diff
@@ -0,0 +1,57 @@
+                vaultDetails = json.loads(subprocess.run(f"op vault get {vaultID} --format=json", shell=True, check=True, capture_output=True).stdout)
+                vaultName = vaultDetails['name']
+            except (Exception) as err:
+                print(f"Encountered an error getting the details for vault with ID {vaultID}: ", err)
```
I might have been poorly Pythoning, but this script errored out seemingly weirdly at this point:

```
Encountered an error getting the details for vault with ID vaultName,vaultUUD,itemCount,vaultCreated,vaultUpdated,userCount:  Command 'op vault get vaultName,vaultUUD,itemCount,vaultCreated,vaultUpdated,userCount --format=json' returned non-zero exit status 1.
Encountered an error getting the details for vault with ID bippity,ddsweci7li5jubu4izcnkfaj4y,0,2023-9-1 22:30:6,2023-9-1 22:30:6,0:  Command 'op vault get bippity,ddsweci7li5jubu4izcnkfaj4y,0,2023-9-1 22:30:6,2023-9-1 22:30:6,0 --format=json' returned non-zero exit status 1.
...
```

scottisloud replied:
Woah that is weird. The exception message definitely suggests something very strange happened. 

did this occur when you ran it with a service account? Did this also happen when you used a 1Password account in the Owners group?

scottisloud replied:
Oh, hang on, I think I understand now. I suspect your list of vault IDs contained additional information? 

The script expects a file like this:

```
usrzroiqicsryaa3o56nddixu4
pyxedwd5d47phwh72l6kyp5h7a
wupizr5o5z4vrehdjsaaicr2cu
```

which I realize is somewhat narrowly defined, but the use-case would be that once you've isolated the candidates for deletion from the csv created by the helper script that you'd copy the uuids into a file that contains only those UUUIds. 

However, would it be better if the script expected a file formatted exactly like the csv? The assumption being the user deleted all rows except the candidates for deletion? 


My only concern about that is it could lead someone to run the script against an _unmodified_ version of the csv file that would delete _all_ vaults as a result. Requiring them to copy and paste the desired UUIDs will force an extra bit of work to prevent a potential mistake. 

I'm on the fence and could easily be swayed one way or another.

ag-adampike replied:
I think I had a different understanding of what the scripts were intended to do, but now that I understand a little better, my two cents is that someone could filter out rows of the CSV output from the helper script based on whatever criteria they liked and then save it with the headers and columns intact for use in _this_ script.

Definitely some documentation will help whichever way you go. :-)

scottisloud replied:
I'd like to provide an off-the-shelf experience that caters to the majority of users the majority of the time. 

Do you think that most script users would want to filter the output at the _script level_, rather than once the csv has been generated? 

Or do you think that most users would most often want to manipulate the CSV using their preferred spreadsheet software?

ag-adampike commented on migration/clean-up-scripts/tests/vault_creator.py:
Did the trick. I also added my own dupes just for fun. :-)

ag-adampike reviewed (commented):
One bit of jankiness with the `add_vault_prefix.py` script for me, but it otherwise seems to be doing the expected job. :-)

scottisloud commented on migration/clean-up-scripts/vault_dedupe_helper.py:
```diff
@@ -0,0 +1,74 @@
+# Get a list of vaults the logged-in user has access to
+def getVaults():
+    try:
+        return subprocess.run("op vault list --group=Owners --format=json", shell=True, check=True, capture_output=True).stdout
```
Since the assumption is that this will be run by an Owner (because it needs to be run by someone with access to all shared vaults), the use of `op vault list --group` is required to ensure that all shared vaults are returned.

`op vault list` will not return vaults an owner only has "manage vault" permissions for, and thus won't produce a comprehensive list of shared vaults since I am assuming that Owners will not have "view items" or greater permissions on all shared vaults.

ag-adampike replied:
I was mistakenly assuming that this script would output a list of vaults that have a name appearing more than once, filtered from a list of all vaults that the authed user/service account has access to. I understand the purpose more clearly now (just get _all_ shared vaults in the account).

scottisloud replied:
That's an important insight, though, Pike. 
* Do you think most people would have the same expectation as you? In other words am I violating a likely-common sense assumption that I need to rethink?
* Regardless of the above, do you think that proper documentation would help? 
* Or should the script behave differently?
* Or should the user be given an option to run the script one way or the other (this is my least favourite approach).

scottisloud commented:
* Added a script that removes the "Imported " prefix from imported vaults. 
* Updated README for cleanup scripts and for the /migration directory to explain each script and how to use them.
* Fixed a potential issue with vault names containing certain special characters in the `add_vault_prefix` script. 

Would appreciate feedback on the documentation:
* Does it make the use-case and usage clearer? 

Would also appreciate feedback on whether my assumptions align with your, and the expectations of "any reasonable person":
* Should the csv report from the dedupe helper list all vaults (current behaviour), or just vaults with identical names? 
* Should I find a way to be more flexible with the format of the input file for the `add_vault_prefix` script? For example, would it make sense to accept the report from `dedupe_helper` and just search for the vaultUUID column? 
  * My fear with this is that someone doesn't remove from the report the vaults they want to keep, and thus delete all vaults in their account when feeding that to the `add_vault_prefix` script in Delete mode.

ag-adampike reviewed (approved):
A couple little suggestions that you can ignore and potentially address in a future PR if you prefer. Happy to approve as is.
